# An assessment of the species diversity and disease potential of Pythium communities in Europe

DOI: https://doi.org/10.1038/s41467-024-52761-0

Authors
---
Wilken Boie*, Markus Schemmel*, Wanzhi Ye, Mario Hasler, Melanie Goll, Joseph-Alexander Verreet and Daguang Cai
<br>
(\* equally contributed to this work)

Author contributions
---
W.B. Performed experiments, data generation, and analysis and wrote the first draft of the paper. M.S. conducted data processing and bioinformatic analysis. W.Y. Performed infection experiments and molecular analysis. M.H. supported for statistical analysis. M.G. developed and supervised the project. JV: developed and supervised the project, and D.C.: designed and supervised experiments and finalized the manuscript. All authors reviewed the manuscript draft. All authors read and approved the final manuscript.

Abstract
---

Pythium sensu lato (s.l.) is a genus of parasitic oomycetes that poses a serious threat to agricultural production worldwide, but their severity is often neglected because little knowledge about them is available. Using an internal transcribed spacer (ITS) amplicon-based-metagenomics approach, we investigate the occurrence, abundance, and diversity of Pythium spp. s.l. in 127 corn fields of 11 European countries from the years 2019 to 2021. We also identify 73 species, with up to 20 species in a single soil sample, and the prevalent species, which show high species diversity, varying disease potential, and are widespread in most countries. Further, we show species-species co-occurrence patterns considering all detected species and link species abundance to soil parameter using the LUCAS topsoil dataset. Infection experiments with recovered isolates show that Pythium s.l. differ in disease potential, and that effective interference with plant hormone networks suppressing JA (jasmonate)-mediated defenses is an essential component of the virulence mechanism of Pythium s.l. species. This study provides a valuable dataset that enables deep insights into the structure and species diversity of Pythium s.l. communities in European corn fields and knowledge for better understanding plantPythium interactions, facilitating the development of an effective strategy to cope with this pathogen.

Description of files and folder structure
---

All input data needed to perform the analysis and produce the figures are provided in the 'Data' folder. Scripts used to perform the sequencing analysis are located in the 'Analysis_sequencing' folder. Custom scripts to create the graphics are in separate folders, as well as for statistical analysis.

The LUCAS 2018 TOPSOIL dataset is not included into the data folder due to licensing restrictions. Please register by your own (https://esdac.jrc.ec.europa.eu/content/lucas-2018-topsoil-data) and download the dataset separately in order to perform the analysis.

| File/Folder | Content |
| ------ | ------- |
|Data    | Input data |
|Analysis_sequencing | Scripts for sequencing analysis |
| Figures | |
|-- Figure_1| Scripts for plotting and data handling |
|-- Figure_2_3| Scripts for plotting and general data export |
|-- Figure_4| Scripts for plotting and statistics |
|-- Figure_5| Scripts for plotting and statistics |
|-- Figure_6| Scripts for plotting |
|-- Figure_7| Scripts for plotting |
|-- Figure_8| Scripts for plotting |
|-- Figure_9| Scripts for plotting and statistics|
|--Figure_10| Scripts for plotting and statistics |
|used_R_packages| List of all R packages with version number used to perform the analysis |

Finalizing graphics
---
Multipanel plots and plots containing images were arranged and finalized using Inkscape (v1.3).
